[C#] New Concept Exercise: arrays (#1257)
exercism/v3: languages/csharp/exercises/concept/arrays/.docs/introduction.md

> In C#, data structures that can hold zero or more elements are known as _collections_. An array is a collection with a fixed size and which elements must all be of the same type. Elements in an array can be assigned to and retrieved by index, with the first element's index being zero:

Reviewer: - Array as bold
- "whose" instead of "which" (we normally personify things in this situation for reasons unknown)
- Reordered the final line to make it parse a little easier.

```suggestion
In C#, data structures that can hold zero or more elements are known as _collections_. An **array** is a collection that has a fixed size and whose contains must all be of the same type. Elements can be assigned to an array or retrieved from it using an index, with the first element's index being zero:
```

```csharp
// Declare array with explicit size (size is 2)
int[] twoIntegers = new int[2];

// Assign first and second element by index
twoIntegers[0] = 7;
twoIntegers[1] = 8;

// Retrieve the second element by index
twoIntegers[1] == 8; // => true
```

Arrays can also be defined using a shortcut notation that allows you to both create the array and set its value. As the compiler can now tell how many elements the array will have, the length can be omitted:

```csharp
// Assign and initialize (size is 3)

> double[] threeDoublesV1 = new double[] { 1.2, 4.45, 9.998 };

Reviewer: I suggest using integers, not floating points in these example. It makes it easier to parse the `.` vs `,`.

Author: Yeah, I wanted to show use a different type, but integers are clearer.

// Shorter notation (size is 3)
int[] threeIntsV2 = new[] { 4, 9, 7 };

// Shortest notation (size is 3)
int[] threeIntsV3 = { 4, 9, 7 };
```

Arrays can be manipulated by either calling an array's methods or properties, or by using the methods defined in the `Array` class.

> Besides access values by index, the fact that an array is also a _collection_ means that you can iterate over all its values using a `foreach` loop:

Reviewer 2: ```suggestion
The fact that an array is also a _collection_ means that besides accessing values by index, you can iterate over all its values using a `foreach` loop:
```

Reviewer: I like the change. I'd like another comma though :)

```suggestion
The fact that an array is also a _collection_ means that, besides accessing values by index, you can iterate over all its values using a `foreach` loop:
```

> In C#, data structures that can hold zero or more elements are known as _collections_. An **array** is a collection that has a fixed size and whose elements must all be of the same type. Elements can be assigned to an array or retrieved from it using an index, with the first element's index being zero:

Reviewer 2: Apologies, I thought I commented this before, but I don't see it on history now.
People transitioning from other languages may be scanning for specific terms (zero-based), and newbies will still need the explanation.

```suggestion
In C#, data structures that can hold zero or more elements are known as _collections_. An **array** is a collection that has a fixed size and whose elements must all be of the same type. Elements can be assigned to an array or retrieved from it using an index. C# arrays are zero-based, meaning that the first element's index is always zero:
```

```csharp
// Declare array with explicit size (size is 2)
int[] twoInts = new int[2];

// Assign first and second element by index
twoInts[0] = 7;
twoInts[1] = 8;

// Retrieve the second element by index
twoInts[1] == 8; // => true

// Check the length of the array
twoInts.Length == 2; // => true
```

Arrays can also be defined using a shortcut notation that allows you to both create the array and set its value. As the compiler can now tell how many elements the array will have, the length can be omitted:

```csharp
// Assign and initialize (size is 3)
int[] threeIntsV1 = new int[] { 4, 9, 7 };

// Shorter notation (size is 3)
int[] threeIntsV2 = new[] { 4, 9, 7 };

// Shortest notation (size is 3)
int[] threeIntsV3 = { 4, 9, 7 };
```

> Arrays can be manipulated by either calling an array's methods or properties, or by using the methods defined in the `Array` class.

Reviewer 3: ```suggestion
Arrays can be manipulated by using the methods and properties defined in the `Array` class.
```

Author: This suggestion is not true. There _are_ properties and instance methods on an array, such as the `Length` property and the `CopyTo()` method, and it is important to know that there are.

The fact that an array has methods becomes especially important later on, when students will know about LINQ's extension methods, which also apply to arrays.

Reviewer 3: And these methods and  properties are defined on the array class right? 
Or you imply on the second half of the sentence that there are static methods on the Array class?

I just think that the sentence is confusing, and it links three times to the array class...  and it looks like the message is just that an array is an object of the Array class, then everything else is implied.

Author: Ah right, I didn't get what you were saying. I've updated the documentation to make the distinction more clear. Sorry for the confusion.
